internal/cmd/cli/apply/apply.go

```diff
@@ -225,26 +239,95 @@ func Dir(ctx context.Context, client Getter, name, baseDir string, opts *Options
 	}
 
 	if opts.Output == nil {
-		err = save(client, def, scans...)
+		err = save(ctx, client, def, opts, scans...)
 	} else {
 		_, err = opts.Output.Write(b)
 	}
 
 	return err
 }
 
-func save(client Getter, bundle *fleet.Bundle, imageScans ...*fleet.ImageScan) error {
+func pushOCIManifest(ctx context.Context, bundle *fleet.Bundle, opts *Options) (string, error) {
+	manifest := manifest.FromBundle(bundle)
+	manifestID, err := manifest.ID()
+	if err != nil {
+		return "", err
+	}
+	ociOpts := ociutils.OCIOpts{
+		URL:             opts.OCIRegistry.URL,
+		Username:        opts.OCIRegistry.Username,
+		Password:        opts.OCIRegistry.Password,
+		BasicHTTP:       opts.OCIRegistry.BasicHTTP,
+		InsecureSkipTLS: opts.OCIRegistry.InsecureSkipTLS,
+	}
+	err = ociutils.PushManifest(ctx, ociOpts, manifestID, manifest)
+	if err != nil {
+		return "", err
+	}
+	return manifestID, nil
+}
+
+func save(ctx context.Context, client Getter, bundle *fleet.Bundle, opts *Options, imageScans ...*fleet.ImageScan) error {
 	c, err := client.Get()
 	if err != nil {
 		return err
 	}
 
 	obj, err := c.Fleet.Bundle().Get(bundle.Namespace, bundle.Name, metav1.GetOptions{})
+	var manifestID string
 	if apierrors.IsNotFound(err) {
-		if _, err = c.Fleet.Bundle().Create(bundle); err != nil {
+		if opts.OCIRegistry.URL != "" {
+			// we have a OCI registry url.
+			// use the OCI registry instead of the Bundle + Content resource approach
+			manifestID, err = pushOCIManifest(ctx, bundle, opts)
+			if err == nil {
+				// OCI artifact stored successfully
+				// We don't store the resources in the bundle. Just keep the manifestID for
+				// being able to access the bundle's contents later.
+				bundle.Spec.Resources = nil
+				bundle.Spec.ContentsID = manifestID
+			} else {
+				return err
+			}
+		}
+		var bundleCreated *fleet.Bundle
+		bundleCreated, err = c.Fleet.Bundle().Create(bundle)
```

Comment:
Because it overwrites the error from: 
```go
obj, err := c.Fleet.Bundle().Get(bundle.Namespace, bundle.Name, metav1.GetOptions{})
```
which is a `NotFound` error. 

It was previously:
```go
obj, err := c.Fleet.Bundle().Get(bundle.Namespace, bundle.Name, metav1.GetOptions{})
	if apierrors.IsNotFound(err) {
		if _, err = c.Fleet.Bundle().Create(bundle); err != nil {
```

But with OCI we can't create the bundle at that point, because we still need to check if we want to use OCI or not.
In the case that we don't want to use OCI it jumps to create the `bundle` the error is not updated and it is returned at the very end (still being a NotFound error).

I know this is confusing, we could also ignore the NotFound error at the end... but I think being so far away from the context were it happened (in the initial Get call) is not ideal.

Another approach would be to clear the err to `nil` right after this:
```go 
obj, err := c.Fleet.Bundle().Get(bundle.Namespace, bundle.Name, metav1.GetOptions{})
	if apierrors.IsNotFound(err) {
```

and then use the `:=` as suggested. 
I changed this to the later.

Reply:
Interesting, thanks for clarifying. This makes me wonder if that `return err` at the very end of the function body makes sense.
My understanding is that it may as well be a `return nil`, as any error would have been handled and have led to an earlier return.
Wouldn't reaching that line mean that any error (eg. a bundle not being found) can be safely ignored?

Reply:
I've changed the function because IMHO it was too large and hard to read.
Now the OCI stuff is done in a separated function. 

I'm also returning nil at the end as suggested

Reply:
Nice one, thanks for taking the time to make this more readable!